sw/device/tests/otbn/otbn_ecdsa_p256_test.c

```diff
@@ -0,0 +1,262 @@
+
+  // Call OTBN to perform operation, and wait for it to complete.
+  CHECK(otbn_call_function(otbn_ctx, kOtbnAppP256EcdsaFuncVerify) == kOtbnOk);
+  CHECK(otbn_busy_wait_for_done(otbn_ctx) == kOtbnOk);
+
+  // Read back results.
+  CHECK(otbn_copy_data_from_otbn(otbn_ctx, /*len_bytes=*/32, kOtbnVarRnd,
+                                 signature_r_out) == kOtbnOk);
+}
+
+/**
+ * Performs a ECDSA roundtrip test using the NIST P-256 curve.
+ *
+ * A roundtrip consists of three steps: Initialize OTBN, sign, and verify.
+ */
+static void test_ecdsa_p256_roundtrip(void) {
+  // message
+  static const uint8_t kIn[32] = {"Hello OTBN."};
+
+  // public key x-coordinate (Q.x)
+  static const uint8_t kPublicKeyQx[32] = {
+      0x4e, 0xb2, 0x8b, 0x55, 0xeb, 0x88, 0x62, 0x24, 0xf2, 0xbf, 0x1b,
+      0x9e, 0xd8, 0x4a, 0x09, 0xa7, 0x86, 0x67, 0x92, 0xcd, 0xca, 0x07,
+      0x5d, 0x07, 0x82, 0xe7, 0x2d, 0xac, 0x31, 0x14, 0x79, 0x1f};
+
+  // public key y-coordinate (Q.y)
+  static const uint8_t kPublicKeyQy[32] = {
+      0x27, 0x9c, 0xe4, 0x23, 0x24, 0x10, 0xa2, 0xfa, 0xbd, 0x53, 0x73,
+      0xf1, 0xa5, 0x08, 0xf0, 0x40, 0x9e, 0xc0, 0x55, 0x21, 0xa4, 0xf0,
+      0x54, 0x59, 0x00, 0x3e, 0x5f, 0x15, 0x3c, 0xc6, 0x4b, 0x87};
+
+  // private key (d)
+  static const uint8_t kPrivateKeyD[32] = {
+      0xcd, 0xb4, 0x57, 0xaf, 0x1c, 0x9f, 0x4c, 0x74, 0x02, 0x0c, 0x7e,
+      0x8b, 0xe9, 0x93, 0x3e, 0x28, 0x0c, 0xf0, 0x18, 0x0d, 0xf4, 0x6c,
+      0x0b, 0xda, 0x7a, 0xbb, 0xe6, 0x8f, 0xb7, 0xa0, 0x45, 0x55};
+
+  // secret random integer (k)
+  static const uint8_t kSecretRandomIntK[32] = {
+      0x7d, 0x07, 0x40, 0x6b, 0x57, 0x4f, 0xc1, 0xfe, 0x0a, 0x56, 0x3e,
+      0x81, 0xd4, 0x10, 0x75, 0x6a, 0x86, 0x8e, 0x1d, 0x07, 0x42, 0x25,
+      0xc9, 0xeb, 0x89, 0xae, 0x7b, 0x02, 0x49, 0x87, 0x65, 0x4d};
+
+  // signature component r
+  static const uint8_t kExpectedSignatureR[32] = {
+      0xb2, 0x9c, 0x5e, 0xc1, 0xff, 0x9c, 0xa9, 0x70, 0xb8, 0xd7, 0x7d,
+      0x90, 0xbb, 0x29, 0x8d, 0x5f, 0x8f, 0x96, 0x42, 0x66, 0x5f, 0xe8,
+      0x49, 0x59, 0xf7, 0x43, 0xb8, 0x03, 0x5b, 0xf7, 0x89, 0x04};
+
+  // signature component s
+  static const uint8_t kExpectedSignatureS[32] = {
+      0xdd, 0x33, 0x60, 0x07, 0x1e, 0x60, 0x06, 0xa4, 0xe2, 0xfa, 0x41,
+      0xf9, 0xf1, 0x69, 0xa5, 0x79, 0x38, 0x0c, 0xcd, 0x66, 0x88, 0xda,
+      0x25, 0x2e, 0x23, 0x56, 0x81, 0x62, 0x28, 0xf8, 0x6b, 0xd1};
+
+  // Initialize
+  otbn_t otbn_ctx;
+  dif_otbn_config_t otbn_config = {
+      .base_addr = mmio_region_from_addr(TOP_EARLGREY_OTBN_BASE_ADDR),
+  };
+  profile_start();
+  CHECK(otbn_init(&otbn_ctx, otbn_config) == kOtbnOk);
+  CHECK(otbn_load_app(&otbn_ctx, kOtbnAppP256Ecdsa) == kOtbnOk);
+  profile_end("Initialization");
+
+  // Sign
+  uint8_t signature_r[32] = {0};
+  uint8_t signature_s[32] = {0};
+
+  LOG_INFO("Signing");
+  profile_start();
+  p256_ecdsa_sign(&otbn_ctx, kIn, kSecretRandomIntK, kPrivateKeyD, signature_r,
+                  signature_s);
```

Comment:
Do we really want to include the `check_data` calls in the timing information for sign operations? If we do then should we print it separately?

Reply:
This also applies to the verify test.

Reply:
Sometimes :-)
* For the sign test we don't need to include it, I pulled the checking out of the profiling.
* For the verify test we want to include the checking, as it is part of the algorithm. I added a comment stating that to the code.